### PR TITLE
Turn screen on Plugin

### PR DESCRIPTION
**Pull request by nazar-kutz**

### CODE REVIEW

**Comment by vshcherb** [comment]
Please rename OsmAnd-Plugin-TurnScreenOn

File: OsmAnd-Plugin-TurnScreenOn/src/main/java/net/osmand/turnScreenOn/listener/MessageSender.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package net.osmand.turnScreenOn.listener;
+
+public interface MessageSender {
```
Comment (crimean) [inline OsmAnd-Plugin-TurnScreenOn/src/main/java/net/osmand/turnScreenOn/listener/MessageSender.java]: Looks like abstract, but used for single purpose. Needs to be renamed?

File: OsmAnd-Plugin-TurnScreenOn/src/main/java/net/osmand/turnScreenOn/helpers/RadioGroupWrapper.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package net.osmand.turnScreenOn.helpers;
+
+import android.util.SparseArray;
+import android.widget.RadioButton;
+
+public class RadioGroupWrapper {
```
Comment (crimean) [inline OsmAnd-Plugin-TurnScreenOn/src/main/java/net/osmand/turnScreenOn/helpers/RadioGroupWrapper.java]: Purpose is unclear.

File: OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java [hunk 3]
```diff
@@ -937,13 +939,21 @@ private void makeSound() {
...
 		voiceMessageListeners.put(voiceMessageListener, 0);
 	}
 	
 	public void removeVoiceMessageListener(VoiceMessageListener voiceMessageListener) {
+		Log.d("ttpl", "remove listener");
 		voiceMessageListeners.remove(voiceMessageListener);
 	}
 
 	public void notifyOnVoiceMessage() {
+		Log.d("ttpl", "notify listeners: ");
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java]: Log.d should be deleted - only for dev and we use commons logging for prod

File: OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java [hunk 4]
```diff
@@ -1897,6 +1899,33 @@ public void unregisterFromUpdates(long id) {
 		navUpdateCallbacks.remove(id);
 	}
 
+	private Map<Long, VoiceRouter.VoiceMessageListener> voiceRouterMessageCallbacks= new ConcurrentHashMap<>();
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java]: Fields in top of the class

File: OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java [hunk 5]
```diff
@@ -937,13 +939,21 @@ private void makeSound() {
...
+		int i = 0;
+		for (VoiceMessageListener lnt : voiceMessageListeners.keySet()) {
+			Log.d("ttpl", "listner " + ++i);
+			lnt.onVoiceMessage();
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java]: Paramater which voice message should be present

File: OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java [hunk 6]
```diff
@@ -8,6 +8,7 @@
 import java.util.concurrent.ConcurrentHashMap;
 
 import net.osmand.Location;
+import net.osmand.aidl.OsmandAidlService;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java]: this class shouldn't be present here

File: OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java [hunk 7]
```diff
@@ -27,6 +28,7 @@
 import alice.tuprolog.Term;
 import android.media.AudioManager;
 import android.media.SoundPool;
+import android.util.Log;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java]: class should be deleted

File: OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java [hunk 8]
```diff
@@ -27,6 +28,7 @@
...
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java]: Use weakreference for listeners

File: OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java [hunk 9]
```diff
@@ -1897,6 +1903,43 @@ public void unregisterFromUpdates(long id) {
...
+			@Override
+			public void onVoiceMessage() {
+				if (aidlCallbackListener != null) {
+					for (OsmandAidlService.AidlCallbackParams cb : aidlCallbackListener.getAidlCallbacks().values()) {
+						if (!aidlCallbackListener.getAidlCallbacks().isEmpty() && (cb.getKey() & KEY_ON_VOICE_MESSAGE) > 0) {
+							try {
+								cb.getCallback().onVoiceRouterNotify();
+							} catch (Exception e) {
+								LOG.debug(e.getMessage(), e);
+							}
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java]: debug?

File: OsmAnd/src/net/osmand/plus/OsmandSettings.java [hunk 10]
```diff
@@ -1387,6 +1387,24 @@ protected Integer getDefaultValue() {
 		KEEP_INFORMING.setModeDefaultValue(ApplicationMode.BICYCLE, 0);
 		KEEP_INFORMING.setModeDefaultValue(ApplicationMode.PEDESTRIAN, 0);
 	}
+	
+	public final CommonPreference<Integer> WAKE_ON_VOICE_TIME_INT = new IntPreference("wake_on_voice_time_int", 0).makeProfile();
+
+	{
+		// 0 means never
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/OsmandSettings.java]: // 0 means never?

File: OsmAnd/src/net/osmand/plus/OsmandSettings.java [hunk 11]
```diff
@@ -1387,6 +1387,24 @@ protected Integer getDefaultValue() {
...
+		WAKE_ON_VOICE_TIME_INT.setModeDefaultValue(ApplicationMode.CAR, 0);
+		WAKE_ON_VOICE_TIME_INT.setModeDefaultValue(ApplicationMode.BICYCLE, 0);
+		WAKE_ON_VOICE_TIME_INT.setModeDefaultValue(ApplicationMode.PEDESTRIAN, 0);
+	}
+
+	public final CommonPreference<Boolean> WAKE_ON_VOICE_SENSOR = new BooleanPreference("wake_on_voice_sensor", false).makeProfile();
+
+	{
+		// 0 means never
+		WAKE_ON_VOICE_SENSOR.setModeDefaultValue(ApplicationMode.CAR, false);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/OsmandSettings.java]: delete theses blocks

File: OsmAnd/src/net/osmand/plus/helpers/LockHelper.java [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+			timedUnlock(screenPowerSave * 1000L);
+		}
+	}
+
+	public void switchSensor(boolean enable) {
+		if (enable) {
+			LOG.debug("switch on sensor");
+			mSensorManager.registerListener(this, mProximity, SensorManager.SENSOR_DELAY_NORMAL);
+		} else {
+			LOG.debug("switch off sensor");
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/helpers/LockHelper.java]: debug?

File: OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java [hunk 13]
```diff
@@ -765,35 +759,35 @@ private void playMakeTurn(RouteSegmentResult currentSegment, RouteDirectionInfo
...
+						p.makeUT(nextNextInfo.distanceTo, new StreetName());
 					}
 				}
 			}
 			if (isplay) {
-				notifyOnVoiceMessage();
-				play.play();
+				p.play();
 			}
+			play(null);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java]: double play?

File: OsmAnd/src/net/osmand/plus/helpers/LockHelper.java [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+		}
+	}
+
+	public void switchOffSensor() {
+		LOG.debug("switch off sensor");
+		mSensorManager.unregisterListener(this);
+	}
+
+	public void switchOnSensor() {
+		LOG.debug("switch on sensor");
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/helpers/LockHelper.java]: delete debug

File: OsmAnd/src/net/osmand/plus/helpers/LockHelper.java [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+
+		mSensorManager = (SensorManager) app.getSystemService(Context.SENSOR_SERVICE);
+		mProximity = mSensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY);
+
+		switchOnSensor();
+
+		voiceMessageListener = new VoiceRouter.VoiceMessageListener() {
+			@Override
+			public void onVoiceMessage() {
+				LOG.debug("onVoiceMessage");
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/helpers/LockHelper.java]: debug?

File: OsmAnd/src/net/osmand/aidl/IOsmAndAidlInterface.aidl [hunk 16]
```diff
@@ -705,4 +706,7 @@ interface IOsmAndAidlInterface {
     boolean areOsmandSettingsCustomized(in OsmandSettingsInfoParams params);
 
     boolean setCustomization(in CustomizationInfoParams params);
+    long registerForVoiceRouterMessages(in ANavigationVoiceRouterMessageParams params, IOsmAndAidlCallback callback);
+    
+    boolean changeMapActivityKeyguardFlags(in boolean enable);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/IOsmAndAidlInterface.aidl]: add comments

File: OsmAnd/src/net/osmand/plus/helpers/LockHelper.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+		voiceMessageListener = new VoiceRouter.VoiceMessageListener() {
+			@Override
+			public void onVoiceMessage() {
+				unlockEvent();
+			}
+		};
+
+		setVoiceRouterListener(true);
+		setSensor(true);
```
Comment (crimean) [inline OsmAnd/src/net/osmand/plus/helpers/LockHelper.java]: Start sensor only if turned on in settings

File: OsmAnd/src/net/osmand/plus/helpers/LockHelper.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+	@Override
+	public void onAccuracyChanged(Sensor sensor, int accuracy) {
+
+	}
+
+	@Override
+	public void onSensorChanged(SensorEvent event) {
+		if (event.sensor.getType() == Sensor.TYPE_PROXIMITY) {
+			if (event.values[0] >= -SENSOR_SENSITIVITY && event.values[0] <= SENSOR_SENSITIVITY) {
+				if (isSensorEnabled()) {
```
Comment (crimean) [inline OsmAnd/src/net/osmand/plus/helpers/LockHelper.java]: Such check must be at top of the method

File: OsmAnd/src/net/osmand/plus/helpers/LockHelper.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+		mProximity = mSensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY);
+
+		voiceMessageListener = new VoiceRouter.VoiceMessageListener() {
+			@Override
+			public void onVoiceMessage() {
+				unlockEvent();
+			}
+		};
+
+		setVoiceRouterListener(true);
```
Comment (crimean) [inline OsmAnd/src/net/osmand/plus/helpers/LockHelper.java]: Why it is not the same - https://github.com/osmandapp/Osmand/commit/f9cd73d4f02b45c429a94e5444124cc6f3463111#diff-33da24854a630e0918b813683c839fbbL79

**Pull request merged**